src/Application/FileProcessor.php

```diff
@@ -50,6 +52,11 @@ public function __construct(
 
     public function processFile(File $file, Configuration $configuration): FileProcessResult
     {
+        if ($configuration->isSecondRun()) {
+            // 2nd run
+            $this->rectorNodeTraverser->prepareCollectorRectorsRun($configuration);
```

Comment:
does it mean, on the "2nd run" we only run CollectorRectors, and on the 1st run, we run regular-rectors and data-collectors side-by-side?

Reply:
since you put a :+1: on it, let me explain further:

I think letting collectors run, while regular rector are still beeing applied and move things arround is wasteful.
wouldn't it make sense to apply all regular rectors first. after the changes stabilized let the collectors run and after that run the collectorrectors (and after that start over)? 

or is this something PHPStan pre-defines and we can't influence?

Reply:
> if so, I am wondering whether regular-rectors which modify the AST can lead to invalid/no longer up 2 date data beeing collected by the collectors

The collectors only collect metadata, not nodes. E.g. in this PR looks for classes, that are extended. These classes cannot be final then.

On 2nd run only the `CollectorRectorInterface` will iterate over nodes and only finalize classes, that are not extended.

Reply:
> wouldn't it make sense to apply all regular rectors first. after the changes stabilized let the collectors run and after that run the collectorrectors (and after that start over)?

I think that's how it's done now:

* 1st run only normal Rectors + collect metadata
* 2nd run use collected metadata and run only Rector rules that will modify nodes based on those metadata

Reply:
> The collectors only collect metadata, not nodes. E.g. in this PR looks for classes, that are extended. These classes cannot be final then.
 

I don't think so. the collectors can return whatever they like .. it might collect AST nodes